ember-style-guides.md

```diff
@@ -1,269 +1,312 @@
+The Ember Style Guide
+---------------------
 
-Contents:
+This Ember style guide recommends best practices so that real-world Ember programmers can write code that can be maintained by other real-world Ember programmers. A style guide that reflects real-world usage gets used, and a style guide that holds to an ideal that has been rejected by the people it is supposed to help risks not getting used at all – no matter how good it is.
 
-- General
-- Controllers;
-- Routes;
-- Components;
-- Order of Declaration;
+The guide is separated into several sections of related rules. Rationale for each the rules is provided unless it's fairly obvious.
 
-##General
+## Table of Contents
 
-- Deconstruct local namespace. Cases such as:
+- [Classes](#classes)
+- [Route Handlers](#route-handlers)
+- [Controllers](#controllers)
+- [Components](#components)
+- [Models](#models)
+- [Mixins](#mixins)
+- [Naming](#naming)
+- [Misc](#misc)
+
+## Classes
+
+- Use a consistent structure in your class definitions.
 
 ```javascript
-import Ember from ‘ember’;
+// import statements go first
+import Ember from 'ember';
+
+// namespace deconstruction
+const { Component } = Ember;
 
+// default export
 export default Ember.Component.extend({
-  title: Ember.computed(‘foo’, function() {
-    //something
-  })
+  // body omitted
 });
 ```
 
-should be:
+- Deconstruct local namespaces. This pattern also serves as a handy tool for developers to know what you actually used in a file.
 
 ```javascript
-import Ember from ‘ember’;
+import Ember from 'ember';
+
+// bad
+export default Ember.Component.extend();
+
+// good - single namespace
+const { Component } = Ember;
+
+export default Component.extend();
+
+// bad
+export default Ember.Component.extend({
+  foo: Ember.computed('bar', function() {
+    // return something...
+  })
+});
 
+// good - newline multiple namespaces
 const {
   Component,
   computed
 } = Ember;
 
 export default Component.extend({
-  title: computed(‘foo’, function() {
-    //something
+  foo: computed('bar', function() {
+    // return something...
   })
 });
 ```
 
-another one:
+- Deconstruct methods with care.
 
 ```javascript
+// bad
 const {
-Component,
-inject
+  Component,
+  moment: { add }
 } = Ember;
 
-const { service } = inject;
+export default Component.extend({
+  session: service();
+});
+
+// good - Ember Core methods can be safely deconstructed
+const {
+  Component,
+  inject: { service }
+} = Ember;
 
 export default Component.extend({
-session: inject.service();
+  session: service();
 });
 ```
 
-Where there is only one import, don’t newline the const declaration:
+## Route Handlers
+
+- Use a consistent structure for route handlers.
 
 ```javascript
-const { Component } = Ember;
+export default Route.extend({
+  // service declarations
 
-export default Component.extend();
+  // lifecycle hooks in order of execution
+  model() {},
+  setupController() {},
+
+  // custom functions
+
+  // actions separated by a newline
+});
```

Comment:
This structure example needs examples where there are just comments. It'd be nice to have a common app context throughout the guide. In models we use a blog post.